File: haskell/src/example/Stereotypes.hs
-- Franck Rasolo: https://gist.github.com/3985593
import Data.Char (toLower)
import Data.List (find, intercalate, isInfixOf, sort)

data Stereotype = Stereotype { name :: String, description :: String, aliases :: [String] }
                  deriving (Show)

stereotypes = [
    Stereotype "australia"  "a person from australia"   ["aussie", "legend"]
  , Stereotype "newzealand" "a person from new zealand" ["kiwi"]
  , Stereotype "preston"    "a person from preston"     ["scally"]
  , Stereotype "liverpool"  "a person from liverpool"   ["scouser", "thief"]
  , Stereotype "manchester" "a person from manchester"  ["manc", "mancunian", "fighter"]
  ]

search :: String -> Either String Stereotype
search term = maybe (Left $ message term) Right (find match stereotypes)
    where match stereotype = isInfixOf (map toLower term) (concat $ flatten stereotype)

message term = "Invalid Stereotype [" ++ term ++ "]. Must be one of [" ++ allowed ++ "]"
    where allowed = intercalate ", " (sort $ concatMap flatten stereotypes)

flatten :: Stereotype -> [String]
flatten s = name s : description s : aliases s

File: scala/src/example/Stereotypes.scala
// Rhys Keepence: git://gist.github.com/3983616.git
package example

object Stereotypes {
  val stereotypes = Seq(australia, newzealand, preston, liverpool, manchester)

  def find(name: String): Either[String, Stereotype] = {
    stereotypes
      .find(stereotype => stereotype.names.exists(name.equalsIgnoreCase))
      .toRight("Invalid Stereotype [%s]. Must be one of [%s]".format(name, allPossibleStereotypes))
  }

  def allPossibleStereotypes = stereotypes.flatMap(s => s.names).sorted.mkString(", ")
}

sealed abstract case

class Stereotype(description: String, aliases: String*) {
  val names = toString :: description :: aliases.toList
}

case object australia extends Stereotype("a person from australia", "aussie", "legend")

case object newzealand extends Stereotype("a person from new zealand", "kiwi")

case object preston extends Stereotype("a person from preston", "scally")

case object liverpool extends Stereotype("a person from liverpool", "scouser", "thief")

case object manchester extends Stereotype("a person from manchester", "manc", "mancunian", "fighter")